modules/sd_hijack_optimizations.py

```diff
@@ -92,6 +94,41 @@ def split_cross_attention_forward(self, x, context=None, mask=None):
 
     return self.to_out(r2)
 
+def _maybe_init(self, x):
+    """
+    Initialize the attention operator, if required We expect the head dimension to be exposed here, meaning that x
+    : B, Head, Length
+    """
+    if self.attention_op is not None:
+        return
+    _, M, K = x.shape
+    try:
+        self.attention_op = xformers.ops.AttentionOpDispatch(
+            dtype=x.dtype,
+            device=x.device,
+            k=K,
+            attn_bias_type=type(None),
+            has_dropout=False,
+            kv_len=M,
+            q_len=M,
+        ).op
+    except NotImplementedError as err:
+        raise NotImplementedError(f"Please install xformers with the flash attention / cutlass components.\n{err}")
+
+def xformers_attention_forward(self, x, context=None, mask=None):
+    h = self.heads
+    q_in = self.to_q(x)
+    context = default(context, x)
+    k_in = self.to_k(context)
+    v_in = self.to_v(context)
+    q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> (b h) n d', h=h), (q_in, k_in, v_in))

```

Comment:
xformers can take the inputs in format BNHD, you can skip the rearrange here (and also the rearrange of the output) - that might improve the performance further

Reply:
I added rearrange because of the _maybe_init function. If it could be repurposed for BNHD it'd be great to just pass BNHD.

Reply:
Also, input before rearrange is batch, seq, head * dim, so we'd need to rearrange anyways to split heads, I think.

Reply:
> Also, input before rearrange is batch, seq, head * dim, so we'd need to rearrange anyways to split heads, I think.

this is just a change in the strides of the tensor, and wouldn't require any data copy, so that should be fine in terms of performance.

> It'd run but return a broken image

Did you still split the head from the embedding when doing that? If your format is `B N (H*D)` this will be interpreted as `(B*H) N D` by xformers and produce the wrong output.
Could you print the following to help me debug things?
```
print("attention_op", self.attention_op)
print("dispatch", xformers.ops.AttentionOpDispatch.from_arguments(q, k, v))
print("dispatch_op", xformers.ops.AttentionOpDispatch.from_arguments(q, k, v).op)
```

Reply:
Problem has been resolved, see

https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/1851

Reply:
The thing is that users shouldn't need to specify `op` (unless very specific cases). xformers should figure out the best Operator to dispatch to depending on the inputs.

Reply:
@danthe3rd 
```
attention_op <class 'xformers.ops.MemoryEfficientAttentionCutlassOp'>
dispatch AttentionOpDispatch(dtype=torch.float16, device=device(type='cuda', index=0), k=40, has_dropout=False, attn_bias_type=<class 'NoneType'>, kv_len=77, q_len=4096, kv=40)
dispatch_op <class 'xformers.ops.MemoryEfficientAttentionCutlassOp'>
```

When I had tested, B * H N D failed, as did B N H D.

Reply:
@danthe3rd Tagging you so that you don't waste your time trying to debug this. The proper way (without op argument and without BH N * D to B * H ND) works now. The exact same code, actually. No idea why it ever failed or what made it work.

Reply:
That's great news :)